examples/keyvalue/main.go

```diff
@@ -0,0 +1,149 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+//go:generate thrift-gen --generateThrift --outputDir gen-go --template github.com/uber/ringpop-go/ringpop.thrift-gen --inputFile keyvalue.thrift
+
+package main
+
+import (
+	"flag"
+	"sync"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/uber-common/bark"
+	"github.com/uber/ringpop-go"
+	"github.com/uber/ringpop-go/discovery/jsonfile"
+	gen "github.com/uber/ringpop-go/examples/keyvalue/gen-go/keyvalue"
+	"github.com/uber/ringpop-go/swim"
+	"github.com/uber/tchannel-go"
+	"github.com/uber/tchannel-go/thrift"
+)
+
+var (
+	hostport = flag.String("listen", "127.0.0.1:3000", "hostport to start service on")
+	hostfile = flag.String("hosts", "./hosts.json", "path to hosts file")
+)
+
+type worker struct {
+	adapter gen.TChanKeyValueService
+
+	memoryLock sync.RWMutex
+	memory     map[string]string
+}
+
+func (w *worker) Set(ctx thrift.Context, key string, value string) error {
+	log.Printf("setting key %q to %q", key, value)
+
+	w.memoryLock.Lock()
+	defer w.memoryLock.Unlock()
+
+	if w.memory == nil {
+		w.memory = make(map[string]string)
+	}
+	w.memory[key] = value
+
+	return nil
+}
+
+func (w *worker) Get(ctx thrift.Context, key string) (string, error) {
+	log.Printf("getting key %q", key)
+
+	w.memoryLock.RLock()
+	defer w.memoryLock.RUnlock()
+
+	value := w.memory[key]
+
+	return value, nil
+}
+
+func (w *worker) GetAll(ctx thrift.Context, keys []string) ([]string, error) {
+	log.Printf("getting keys: %v", keys)
+
+	m := make([]string, 0, len(keys))
+	for _, key := range keys {
+		// sharded self call
+		value, err := w.adapter.Get(ctx, key)
+		if err != nil {
+			return nil, err
+		}
+		m = append(m, value)
+	}
+	return m, nil
+}
+
+func main() {
+	flag.Parse()
+
+	ch, err := tchannel.NewChannel("keyvalue", nil)
+	if err != nil {
+		log.Fatalf("channel did not create successfully: %v", err)
+	}
+
+	logger := log.StandardLogger()
+
+	rp, err := ringpop.New("keyvalue",
+		ringpop.Channel(ch),
+		ringpop.Logger(bark.NewLoggerFromLogrus(logger)),
+	)
+	if err != nil {
+		log.Fatalf("Unable to create Ringpop: %v", err)
+	}
+
+	worker := &worker{}
+	adapter, _ := gen.NewRingpopKeyValueServiceAdapter(worker, rp, ch,
+		gen.KeyValueServiceConfiguration{
+			Get: &gen.KeyValueServiceGetConfiguration{
+				Key: func(ctx thrift.Context, key string) (shardKey string, err error) {
+					return key, nil
+				},
+			},
+
+			GetAll: &gen.KeyValueServiceGetAllConfiguration{
+				Key: func(ctx thrift.Context, keys []string) (shardKey string, err error) {
+					// use the node listening on 127.0.0.1:3004 as the node that should answer
```

Comment:
By knowing which node is answering we can craft both a call directly to the node and a call specifically NOT to the node. Otherwise we should call every node from the test to be sure we have all combinations covered.